### PR TITLE
AND-3

### PR DESCRIPTION
**Pull request by lpetka**

### CODE REVIEW

File: setup_environment.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+sudo echo "export JAVA_HOME=/usr/lib/jvm/jdk1.8.0" >> /etc/profile.d/oraclejdk.sh
+sudo echo "export JAVA7_HOME=/usr/lib/jvm/jdk1.7.0" >> /etc/profile.d/oraclejdk.sh
+sudo echo "export JAVA8_HOME=/usr/lib/jvm/jdk1.8.0" >> /etc/profile.d/oraclejdk.sh
+export PATH=$PATH:/usr/lib/jvm/jdk1.8.0/bin:/usr/lib/jvm/jdk1.8.0/bin
+export JAVA_HOME=/usr/lib/jvm/jdk1.8.0
+export JAVA7_HOME=/usr/lib/jvm/jdk1.7.0
+export JAVA8_HOME=/usr/lib/jvm/jdk1.8.0
+
+# download, install and configure android sdk
+cd /tmp/
```
Comment (grzanka) [inline setup_environment.sh]: Why ? Aren't we already in `/tmp`

File: setup_environment.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+sudo mkdir -p /opt/android-sdk/
```
Comment (grzanka) [inline setup_environment.sh]: Do we really need to put Android SDK in system directories ?
Good habit is to use user space wherever possible. Couldn't this SDK go to `$HOME/usr` ?

File: setup_environment.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+# The purpose of this script is to download, install
+# and configure all files and programs required to
+# take part in developing ActivitySync (Android only).
+###################################################################
+
+CURRENT_USER=$USER
+
+# update package lists from repositories
+sudo apt-get update
+sudo apt-get install git
```
Comment (grzanka) [inline setup_environment.sh]: git might be already installed. Better to check if present (sudo not needed), only if missing, install it using sudo.

File: setup_environment.sh [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+sudo mkdir /usr/lib/jvm
```
Comment (grzanka) [inline setup_environment.sh]: the same as above. Check first if java is present and if it is Oracle java. If present, then no need to install it again.

File: setup_environment.sh [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+cd /tmp/
+
+# download, install and configure java7 jdk
+sudo wget --no-check-certificate --no-cookies --header "Cookie: oraclelicense=accept-securebackup-cookie" http://download.oracle.com/otn-pub/java/jdk/7u79-b15/jdk-7u79-linux-x64.tar.gz
+sudo tar -xvf jdk-7*
+sudo mv ./jdk1.7* /usr/lib/jvm/jdk1.7.0
```
Comment (grzanka) [inline setup_environment.sh]: Is it really 1.7.0 ? If it is an unknown version from 1.7 family, you could use 1.7.x

File: setup_environment.sh [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+# update package lists from repositories
+sudo apt-get update
+sudo apt-get install git
+
+sudo mkdir /usr/lib/jvm
+cd /tmp/
+
+# download, install and configure java7 jdk
+sudo wget --no-check-certificate --no-cookies --header "Cookie: oraclelicense=accept-securebackup-cookie" http://download.oracle.com/otn-pub/java/jdk/7u79-b15/jdk-7u79-linux-x64.tar.gz
```
Comment (grzanka) [inline setup_environment.sh]: What if this script will be repeated for a second time in some time (~6 months) from now ?
Will oracle remove JDK 7u79 at this point ?
Does this script require constant maintenance or not ?
Does Oracle provide `-latest` version always pointing to newest JDK ?

File: setup_environment.sh [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+sudo touch /etc/profile.d/android.sh
+sudo chmod a+rwx /etc/profile.d/android.sh
+
+sudo echo '#!/bin/bash' >> /etc/profile.d/android.sh
+sudo echo "export ANDROID_HOME=/opt/android-sdk/" >> /etc/profile.d/android.sh
+sudo echo "export PATH=\$PATH:/opt/android-sdk/tools" >> /etc/profile.d/android.sh
+export ANDROID_HOME=/opt/android-sdk/
+export PATH=$PATH:/opt/android-sdk/tools
+
+sudo chown -R $CURRENT_USER /opt/android-sdk/
```
Comment (grzanka) [inline setup_environment.sh]: Why ?

File: setup_environment.sh [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+android update sdk --no-ui
+sudo apt-get install libstdc++6:i386 libgcc1:i386 zlib1g:i386 libncurses5:i386
```
Comment (grzanka) [inline setup_environment.sh:99]: Why 32-bit libraries are needed ?

File: setup_environment.sh [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+sudo chown -R $CURRENT_USER /home/$CURRENT_USER/.android/
+
+echo "[OK] Rebooting in 3 seconds..."
+sleep 3s
+sudo reboot
```
Comment (grzanka) [inline setup_environment.sh]: Why ???????

File: setup_environment.sh [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+sudo echo "export ANDROID_HOME=/opt/android-sdk/" >> /etc/profile.d/android.sh
+sudo echo "export PATH=\$PATH:/opt/android-sdk/tools" >> /etc/profile.d/android.sh
+export ANDROID_HOME=/opt/android-sdk/
+export PATH=$PATH:/opt/android-sdk/tools
+
+sudo chown -R $CURRENT_USER /opt/android-sdk/
+
+android update sdk --no-ui
+sudo apt-get install libstdc++6:i386 libgcc1:i386 zlib1g:i386 libncurses5:i386
+sudo chown -R $CURRENT_USER /home/$CURRENT_USER/.android/
```
Comment (grzanka) [inline setup_environment.sh]: Why ?

File: setup_environment.sh [hunk 11]
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+###################################################################
+# The purpose of this script is to download, install
+# and configure all files and programs required to
+# take part in developing ActivitySync (Android only).
+###################################################################
+
+CURRENT_USER=$USER
```
Comment (grzanka) [inline setup_environment.sh]: Why not simply using $USER ?

File: setup_environment.sh [hunk 12]
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
```
Comment (grzanka) [inline setup_environment.sh:1]: This script is well written and can be also used to install enviroment on "unclean" system by installing only missing stuff.
In this way development can be done on machines shared with other users (i.e. my wife won't complained that I broken her Java enviroment ;) ).

File: setup_environment.sh (hunk 1 above)
Comment (lpetka) [reply setup_environment.sh]: > Why ? Aren't we already in /tmp

Yes, we are. Will fix.

File: setup_environment.sh (hunk 2 above)
Comment (lpetka) [reply setup_environment.sh]: > Do we really need to put Android SDK in system directories ?
> Good habit is to use user space wherever possible. Couldn't this SDK go to $HOME/usr ?

Good point, will fix.

File: setup_environment.sh (hunk 10 above)
Comment (lpetka) [reply setup_environment.sh]: > Why?

The owner of this directory is root (honestly I have no idea why). Probably it's the effect of this command:
`android update sdk --no-ui`

When user tries to build a project he is told there was no Android SDK because he is not able to access this directory.

File: setup_environment.sh (hunk 11 above)
Comment (lpetka) [reply setup_environment.sh]: > Why not simply using $USER ?

During execution of commands as root (sudo) $USER env variable returns root.

File: setup_environment.sh (hunk 9 above)
Comment (lpetka) [reply setup_environment.sh]: Exports were not working properly for me (I don't know why, I tried many times) so I decided to reboot system to let *.sh files in /etc/profile.d/ take effect. I will remove direct exports from script and let android.sh and oraclejdk.sh do theirs job.

Comment (grzanka) [reply setup_environment.sh]: Logging in and out would be enough. Figureout difference between .bash_profile and .bashrc.

File: setup_environment.sh (hunk 10 above)
Comment (grzanka) [reply setup_environment.sh]: Try to install adroid sdk in user directory and check if this issue still persists

File: setup_environment.sh (hunk 7 above)
Comment (lpetka) [reply setup_environment.sh]: > Why?

I will put Android SDK in user's home directory. Chown will not be needed anymore since the directory will be created with current user's (not root) context.

File: setup_environment.sh (hunk 8 above)
Comment (lpetka) [reply setup_environment.sh:99]: > Why 32-bit libraries are needed ?

This is the fix I found for adb and aapt not working. 
[FIX](http://stackoverflow.com/questions/19523502/how-to-make-androids-aapt-and-adb-work-on-64-bit-ubuntu-without-ia32-libs-work)

Comment (grzanka) [reply setup_environment.sh:99]: Does it mean that Android SDK for 64bit Linux needs to have 32bit libraries. In fact it might be right:
https://developer.android.com/studio/index.html
_(64-bit distribution capable of running 32-bit applications)_

Does it mean that great Google cannot ship code with requiring developers to do strange things ?

File: setup_environment.sh (hunk 3 above)
Comment (lpetka) [reply setup_environment.sh]: > git might be already installed. Better to check if present (sudo not needed), only if missing, install it using sudo.

Will fix.

File: setup_environment.sh (hunk 4 above)
Comment (lpetka) [reply setup_environment.sh]: > the same as above. Check first if java is present and if it is Oracle java. If present, then no need to install it again.

Will fix.

File: setup_environment.sh (hunk 5 above)
Comment (lpetka) [reply setup_environment.sh]: > Is it really 1.7.0 ? If it is an unknown version from 1.7 family, you could use 1.7.x

Will change directory's name to the exact number.

File: setup_environment.sh (hunk 11 above)
Comment (grzanka) [reply setup_environment.sh]: Really ? I might be wrong, but in my case it works a bit differently. Try yourself.

```
# grzanka at gajewski-dell in ~ [12:56:04]
→ echo $USER     
grzanka

# grzanka at gajewski-dell in ~ [12:56:06]
→ sudo echo $USER
grzanka

```

File: setup_environment.sh (hunk 6 above)
Comment (lpetka) [reply setup_environment.sh]: > What if this script will be repeated for a second time in some time (~6 months) from now ?
> Will oracle remove JDK 7u79 at this point ?

JDK 7u79 is the last jdk7 version provided by Oracle. I don't know whether it will be removed or not (but I'm pretty sure it will last at least a few years on their servers).

> Does this script require constant maintenance or not ?

Yes, it does. Script should be checked (and updated if necessary) at least quarterly.

> Does Oracle provide -latest version always pointing to newest JDK ?

I couldn't find any way of updating jdk to latest version but using [ppa:webupd8team/java](http://askubuntu.com/questions/183867/how-do-i-update-oracle-java-7-jdk-and-jre/183868) repository. As you mentioned earlier on trello, we shouldn't use software from unknown sources.

Comment (grzanka) [reply setup_environment.sh]: Can you do simple check and figure out if previously released JDK (7u78 ?) is still present on Oracle servers ?

File: setup_environment.sh (hunk 8 above)
Comment (lpetka) [reply setup_environment.sh:99]: Every solution to this problem (aapt wasn't found at the directory it was located) I found pointed exaclty to install 32-bit libraries. It seemed strange for me but I tried it and it worked so I considered it done.

File: setup_environment.sh (hunk 6 above)
Comment (lpetka) [reply setup_environment.sh]: > Can you do simple check and figure out if previously released JDK (7u78 ?) is still present on Oracle servers ?

[Download archives](http://www.oracle.com/technetwork/java/javase/downloads/java-archive-downloads-javase7-521261.html)

File: setup_environment.sh (hunk 8 above)
Comment (grzanka) [reply setup_environment.sh:99]: ok

File: setup_environment.sh (hunk 11 above)
Comment (lpetka) [reply setup_environment.sh]: Currently I'm not at home. I will check it later, but I remember I struggled with changing ownership using $USER environment variable.

File: setup_environment.sh (hunk 4 above)
Comment (lpetka) [reply setup_environment.sh]: Ok, I have no clue how to check versions of java installed on user's machine.
1. Using `java -version` returns version of jdk set in PATH - gives me nothing, e.g.:
   
   ```
   >java -version
   java version "1.8.0_102"
   Java(TM) SE Runtime Environment (build 1.8.0_102-b14)
   Java HotSpot(TM) 64-Bit Server VM (build 25.102-b14, mixed mode)
   ```
   
   I still know nothing about users jdk7 and vice versa in case it returns jdk7.
2. Using grep to find out wchich versions of jdk user has is also very poor idea because directory names of java may differ, e.g.:
   Installing java from WebUpd8 PPA creates directory _/usr/lib/jvm/java-8-oracle_.
   On the other hand installing java in the way it's done in this script creates directory _/usr/lib/jvm/jdk{version}_
3. Using `update-alternatives --config java` is as bad idea as greping because it uses directories name for identification.

And at this point I run out of ideas.

This is why I put _clean Ubuntu installation_ requirement in my guide. I'm not able to write robust script not knowing user's environment, especially when he is not using package manager.

All that comes to my mind is simple prompt asking for users permission to install and configure given java version.

File: setup_environment.sh [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+export PATH=$PATH:/home/$USER/android-sdk/tools
+
+sudo chown -R $USER /home/$USER/android-sdk/
+
+android update sdk --no-ui
+sudo apt-get install libstdc++6:i386 libgcc1:i386 zlib1g:i386 libncurses5:i386
+
+printf "${GREEN}[OK]${NOCOLOR} Log out in 3 seconds\n"
+sleep 3
+pkill -u $USER
```
Comment (grzanka) [inline setup_environment.sh:104]: Write a comment why pkill is used

File: setup_environment.sh [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+sudo touch /etc/profile.d/android.sh
+sudo chmod a+rwx /etc/profile.d/android.sh
+
+sudo echo '#!/bin/bash' >> /etc/profile.d/android.sh
+sudo echo "export ANDROID_HOME=/home/${USER}/android-sdk/" >> /etc/profile.d/android.sh
+sudo echo "export PATH=\$PATH:/home/${USER}/android-sdk/tools" >> /etc/profile.d/android.sh
+export ANDROID_HOME=/home/$USER/android-sdk/
+export PATH=$PATH:/home/$USER/android-sdk/tools
+
+sudo chown -R $USER /home/$USER/android-sdk/
```
Comment (grzanka) [inline setup_environment.sh]: Why is that needed ?

File: setup_environment.sh [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+sudo echo "export JAVA_HOME=/usr/lib/jvm/jdk1.8.0_102" >> /etc/profile.d/oraclejdk.sh
+sudo echo "export JAVA7_HOME=/usr/lib/jvm/jdk1.7.0_79" >> /etc/profile.d/oraclejdk.sh
+sudo echo "export JAVA8_HOME=/usr/lib/jvm/jdk1.8.0_102" >> /etc/profile.d/oraclejdk.sh
+export PATH=$PATH:/usr/lib/jvm/jdk1.8.0_102/bin
+export JAVA_HOME=/usr/lib/jvm/jdk1.8.0_102
+export JAVA7_HOME=/usr/lib/jvm/jdk1.7.0_79
+export JAVA8_HOME=/usr/lib/jvm/jdk1.8.0_102
+
+# download, install and configure android sdk
+mkdir -p /home/$USER/android-sdk/
```
Comment (grzanka) [inline setup_environment.sh]: Couldn't you use $HOME variable ? Some users have home directories in different location.

File: setup_environment.sh [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+fi
+
+# keep all work inside /tmp directory
+cd /tmp/
+
+# create directory for java
+sudo mkdir /usr/lib/jvm
+
+# download, install and configure java7 jdk
+sudo wget --no-check-certificate --no-cookies --header "Cookie: oraclelicense=accept-securebackup-cookie" http://download.oracle.com/otn-pub/java/jdk/7u79-b15/jdk-7u79-linux-x64.tar.gz
```
Comment (grzanka) [inline setup_environment.sh]: I do not understand. Why root access is needed to get a file using wget and store it in /tmp ?

File: setup_environment.sh [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+sudo tar -xvf jdk-7*
```
Comment (grzanka) [inline setup_environment.sh]: Why sudo ? We are in /tmp, everybody should have the rights to unpack here.

**Comment by grzanka** [comment]
General remark. Do not use sudo when not needed

**Comment by grzanka** [comment]
Why tests are failing ?

**Comment by grzanka** [comment]
A recommendation: in parallel (or in sequence, as you prefer) to making APK on circle CI, you could run the tests, checking if this scripts works fine.

For that purpose pure Ubuntu docker image can be used.

You mentioned before that the script would require maintenance. This would make it easier.

File: setup_environment.sh [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+cd android-sdk-linux/
+sudo mv * /home/$USER/android-sdk/
+
+sudo touch /etc/profile.d/android.sh
+sudo chmod a+rwx /etc/profile.d/android.sh
+
+sudo echo '#!/bin/bash' >> /etc/profile.d/android.sh
+sudo echo "export ANDROID_HOME=/home/${USER}/android-sdk/" >> /etc/profile.d/android.sh
+sudo echo "export PATH=\$PATH:/home/${USER}/android-sdk/tools" >> /etc/profile.d/android.sh
```
Comment (grzanka) [inline setup_environment.sh]: Why are you so brutal. Couldn't these changes be done on user level (using .bashrc or .bash_profile files) ?

I'd even say that better practice is to craft a setup.sh (or setup-activityteam-development.sh) script and put it somewhere, i.e. in android-sdk or somewhere else. In that script you put all modifications of env variables.

Then before each session of development user does: `source somepath/setup.sh` and all is setup.

File: setup_environment.sh (hunk 4 above)
Comment (grzanka) [reply setup_environment.sh]: OK, I agree

File: setup_environment.sh (hunk 14 above)
Comment (grzanka) [reply setup_environment.sh]: You said before:

> I will put Android SDK in user's home directory. Chown will not be needed anymore since the directory will be created with current user's (not root) context

https://github.com/ActivitySyncTeam/Android-ActivitySync/pull/4#discussion_r76394520

Comment (lpetka) [reply setup_environment.sh]: After unpacking android-sdk_r24\* user has no rights to execute files in this directory, e.g. I couldn't run /tools/android later in script. The same principle applies to every executable file download via Internet - before executing file there's a need to set respective rights to allow it.

File: setup_environment.sh (hunk 15 above)
Comment (lpetka) [reply setup_environment.sh]: WIll fix.

File: setup_environment.sh (hunk 13 above)
Comment (lpetka) [reply setup_environment.sh:104]: Will do.

**Comment by lpetka** [comment]
Tests fail because there are no tests written on this branch.
Will remove needless sudo

File: setup_environment.sh (hunk 4 above)
Comment (lpetka) [reply setup_environment.sh]: Will add prompt.

File: setup_environment.sh (hunk 14 above)
Comment (grzanka) [reply setup_environment.sh]: You say:

> After unpacking android-sdk_r24\* user has no rights to execute files in this directory, e.g. I couldn't run /tools/android later in script

If the file `android-sdk_r24*` is owned by the user and unpacked in the directory from which access rights are not inherited then I must say - you are wrong (see code snippet below).

There is also substantial difference between changing owner via `chown` (still I don't get why do you change ownership, who was the owner previously?) and changing access rights via `chmod` (but still after unpacking access rights are correct, see code below).

Note missing `v` in `tar` command.

```
# grzanka at gajewski-dell in ~/tmp [15:50:37]
→ wget https://dl.google.com/android/android-sdk_r24.4.1-linux.tgz
--2016-08-27 15:50:38--  https://dl.google.com/android/android-sdk_r24.4.1-linux.tgz
Resolving dl.google.com (dl.google.com)... 172.217.21.174, 2a00:1450:4001:81d::200e
Connecting to dl.google.com (dl.google.com)|172.217.21.174|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 326412652 (311M) [application/x-tar]
Saving to: ‘android-sdk_r24.4.1-linux.tgz’

100%[=========================================================================>] 326.412.652 5,38MB/s   in 62s    

2016-08-27 15:51:41 (5,02 MB/s) - ‘android-sdk_r24.4.1-linux.tgz’ saved [326412652/326412652]


# grzanka at gajewski-dell in ~/tmp [15:51:41]
→ tar -xzf android-sdk_r24.4.1-linux.tgz 

# grzanka at gajewski-dell in ~/tmp [15:52:02]
→ cd android-sdk-linux/

# grzanka at gajewski-dell in ~/tmp/android-sdk-linux [15:52:07]
→ mkdir -p $HOME/android-sdk

# grzanka at gajewski-dell in ~/tmp/android-sdk-linux [15:52:24]
→ mv * $HOME/android-sdk/

# grzanka at gajewski-dell in ~/tmp/android-sdk-linux [15:52:36]
→ cd $HOME/android-sdk/

# grzanka at gajewski-dell in ~/android-sdk [15:53:20]
→ ls -alh tools/android 
-rwxr--r-- 1 grzanka grzanka 3,5K paź 14  2015 tools/android

# grzanka at gajewski-dell in ~/android-sdk [15:53:31]
→ file tools/android
tools/android: POSIX shell script, ASCII text executable

# grzanka at gajewski-dell in ~/android-sdk [15:53:35]
→ ./tools/android --help

       Usage:
       android [global options] action [action options]
       Global options:
  -s --silent     : Silent mode, shows errors only.
  -v --verbose    : Verbose mode, shows errors, warnings and all messages.
     --clear-cache: Clear the SDK Manager repository manifest cache.
  -h --help       : Help on a specific command.

                                                                    Valid
                                                                    actions
                                                                    are
                                                                    composed
                                                                    of a verb
                                                                    and an
                                                                    optional
                                                                    direct
                                                                    object:
-    sdk              : Displays the SDK Manager window.
-    avd              : Displays the AVD Manager window.
-   list              : Lists existing targets or virtual devices.
-   list avd          : Lists existing Android Virtual Devices.
-   list target       : Lists existing targets.
-   list device       : Lists existing devices.
-   list sdk          : Lists remote SDK repository.
- create avd          : Creates a new Android Virtual Device.
-   move avd          : Moves or renames an Android Virtual Device.
- delete avd          : Deletes an Android Virtual Device.
- update avd          : Updates an Android Virtual Device to match the folders
                        of a new SDK.
- create project      : Creates a new Android project.
- update project      : Updates an Android project (must already have an
                        AndroidManifest.xml).
- create test-project : Creates a new Android project for a test package.
- update test-project : Updates the Android project for a test package (must
                        already have an AndroidManifest.xml).
- create lib-project  : Creates a new Android library project.
- update lib-project  : Updates an Android library project (must already have
                        an AndroidManifest.xml).
- create uitest-project: Creates a new UI test project.
- update adb          : Updates adb to support the USB devices declared in the
                        SDK add-ons.
- update sdk          : Updates the SDK by suggesting new platforms to install
                        if available.


```

Comment (lpetka) [reply setup_environment.sh]: Removing needless sudo should do the trick then :) Thank you.

File: setup_environment.sh (hunk 18 above)
Comment (lpetka) [reply setup_environment.sh]: Will do.

File: setup_environment.sh [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+# download and update essential sdk files
+android update sdk --no-ui
+
+# libraries needed by aapt and adb
+sudo apt-get install libstdc++6:i386 libgcc1:i386 zlib1g:i386 libncurses5:i386
+
+# kill current user session in order to run setup-activityteam-development.sh
+printf "${GREEN}[OK]${NOCOLOR} Log out in 3 seconds.\n"
+sleep 3s
+pkill -u $USER
```
Comment (grzanka) [inline setup_environment.sh:104]: Is it really necessary to kill user session ? 
In my opinion more gentle approach is not to hardcode any PATH modification in .bashrc and do it "on demand".
By "on demand" I mean asking user to run `source $HOME/android-sdk/setup-activityteam-development.sh` every time before starting project development in given terminal.

In such case user could use his machine to develop other projects, requiring different enviroment.

**Comment by grzanka** [comment]
What is the script exit code in case of failure (i.e. lack of internet connection) ?
Hint: `set -e` bash command might help.

**Comment by lpetka** [comment]
>  A recommendation: in parallel (or in sequence, as you prefer) to making APK on circle CI, you could run the tests, checking if this scripts works fine.
> For that purpose pure Ubuntu docker image can be used.
> You mentioned before that the script would require maintenance. This would make it easier.

Honestly, I have no idea how these tests would look like. In this particular case maintenance includes only updating download links for JDK8 (Android SDK download updates when needed during app build and there wouldn't be newer JDK7). Therefore I don't see any point in mantaining server with Ubuntu and configuring docker just for this one little thing. In my opinion it's just overkill.

**Comment by grzanka** [comment]
There is no need of `mantaining server with Ubuntu`

CircleCI is based on docker, so this test in its most basic form would be a single line in .circleci.yml running sth like:   - `docker run ....setup_environment.sh`
Then I assume that exit codes in setup_environment.sh are handled correctly.
We could then assume that if exit code 0 was given then setup was successful.

What is the advantage of that approach is that having a docker instance correctly configured with setup_environment.sh one could run also `./gradlew assemble` to check if building was correct.

This would ensure us that:
- setup_environment.sh correctly configures the environment for development
- gradle correctly builds in the environment configured using abovementioned script

Successful build in gradle doesn't mean successful build in the enviroment configured using setup_environment.sh script.

Take it as advice only.

File: circle.yml [hunk 20]
```diff
@@ -0,0 +1,9 @@
+machine:
+  services:
+    - docker
+
+test:
+    override:
+        - (chmod +x setup_environment.sh)
```
Comment (grzanka) [inline circle.yml:7]: Not necessary. I am assuming you work on Linux.
Before doing (git add + git commit) do locally on your PC: `chmod +x setup_environment.sh`
GIT keeps the file together with mode of the file.

**Comment by grzanka** [comment]
I see here multiple tests of docker. I recommend to do it locally on own PC instead of making trial-and-error commits.
I suggest also to investigate exit code of the script. The whole purpose of the automatic testing is to mark test as failing if sth wrong happens. Try to check what is the exit code of the script (`$?`) if it contains some line with non-zero exit code (like `ls frodo baggins` ;). Does docker run also gives non-zero exit code in that case ? Bash command `set -e` could help - check what is it doing.

**Comment by grzanka** [comment]
Check this one:
https://github.com/pypa/python-manylinux-demo/blob/master/.travis.yml
The key is `-v` option (read first what is it doing).

Do you know how does docker works ? Does it execute your script as root or as an user ?
